packages/swingset-liveslots/src/boyd-gc.js

```diff
@@ -0,0 +1,438 @@
+  facet gets is own vref, but they all share a common prefix, known as
+  the "baseRef". For example, `o+d44/2` is a BaseRef for a cohort, the
+  second instance created of the `o+d44` Kind, whose individual facets
+  would have vrefs of `o+d44/2:0` and `o+d44/2:1`.
+
+  We use a WeakMap to ensure that holding any facet will keep all the
+  others alive, so the cohort lives and dies as a group. The GC
+  tracking code needs to track the whole cohort at once, not the
+  individual facets, and any data structure which refers to cohorts
+  will use the BaseRef instead of a single vref. So `slotToVal` is
+  keyed by a BaseRef, and its values are a cohort of facets. But
+  `valToSlot` is keyed by the facets, and its values are the
+  individual facet's vref.
+
+  Most of the GC-related APIs that appear here take vrefs, but the
+  exceptions are:
+  * slotToVal is keyed by baseRef
+  * possiblyDeadSet holds baseRefs, that's what our WeakRefs track
+  * vrm.isVirtualObjectReachable takes baseRef
+  * vrm.deleteVirtualObject takes baseRef, returns [bool, retireees=vrefs]
+  * vrm.ceaseRecognition takes either baseRef or vref
+    (if given a baseRef, it will process all the facets)
+
+  For Presence- and Remotable- style objects, the vref and baseref are
+  the same.
+*/
+
+export const makeBOYDKit = ({
+  gcTools,
+  slotToVal,
+  vrm,
+  kernelRecognizableRemotables,
+  syscall,
+  possiblyDeadSet,
+  possiblyRetiredSet,
+}) => {
+  // Presence (o-NN) lifetimes are controlled by the upstream vat, or
+  // the kernel. If the vref was in possiblyDeadSet, then it *was*
+  // reachable before, so we can safely presume the kernel to think we
+  // can reach it.
+
+  const checkImportPresence = vref => {
+    // RAM pillar || vdata pillar
+    const isReachable = slotToVal.has(vref) || vrm.isPresenceReachable(vref);
+    // use slotToVal.has, not getSlotForVal(), to avoid duplicate drop
+    let dropImport;
+    if (!isReachable) {
+      dropImport = vref;
+    }
+    return { dropImport };
+  };
+
+  // Remotable (o+NN) lifetimes are controlled by us: we delete/retire
+  // the object as soon as it becomes unreachable. We only track the
+  // Remotable/Far object (the RAM pillar) directly: exports retain
+  // the Remotable in the exportedRemotables set, and vdata retains it
+  // as keys of the remotableRefCounts map. So when we decide the vref
+  // is unreachable, the Remotable is already gone, and it had no
+  // other data we need to delete, so our task is to remove any local
+  // recognition records, and to inform the kernel with a
+  // retireExports if kernelRecognizableRemotables says that the
+  // kernel still cares.
+
+  // note: we track export status for remotables in the
+  // kernelRecognizableRemotables set, not vom.es.VREF records. We
+  // don't currently track recognition records with
+  // vom.ir.VREF|COLLECTION, but we should, see #9956
+
+  const checkExportRemotable = vref => {
```

Comment:
Visual grouping suggestion:
```suggestion
  // recognition records, and to inform the kernel with a
  // retireExports if kernelRecognizableRemotables says that the
  // kernel still cares.
  //
  // note: we track export status for remotables in the
  // kernelRecognizableRemotables set, not vom.es.VREF records. We
  // don't currently track recognition records with
  // vom.ir.VREF|COLLECTION, but we should, see #9956
  const checkExportRemotable = vref => {
```

Reply:
I like the extra `//`, but removing the newline at the end makes the start of the function less obvious on my screen.. I'll apply half